### PR TITLE
Use a PR for Release Commits

### LINKED ISSUE #505: Replace usage of ADMIN_GITHUB_TOKEN
### Problem

We should not have to worry about the ADMIN_GITHUB_TOKEN.  It is a maintenance burden and an additional security surface area.

### Proposed Solution

We should instead create a PR with the commits that were previously pushed directly, and merge it from the Action.  This will allow branch protections to remain in place.  We can recommend the use of `environments` with restricted users in the publish action, regardless of whether OIDC tokens are used (#504).

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #505

- Remove the need for `ADMIN_GITHUB_TOKEN`
- Create a PR and merge it to make the actual changes in populate release
- Create and push the tag as part of finalize release
- Steps:
  - [x] Test this using https://github.com/blink1073/test-python-project
  - [x] Get tests to pass
  - [x] Add notes on how to test changes - use the fork and branch as the `@`
  - [x] Update examples and docs - must always use environment for security - should use trusted publishers
  - [x] Publish this as the v3 tag
  - [x] Update docs and example workflows for the new fine-grained token access

### CODE REVIEW

File: jupyter_releaser/lib.py
Comment (fcollonval) [inline jupyter_releaser/lib.py:165]: Typo

```suggestion
    util.log('Creating a PR')
```

File: jupyter_releaser/lib.py
Comment (fcollonval) [inline jupyter_releaser/lib.py:186]: If I follow correctly, this will use the GH API to merge the release PR. If it fails the populate release step will fail. Is that correct?

---

What happen if pull requests have protection rule; e.g. on JupyterLab it is required that the linter job succeed. Will this line fail in that case?

**Review by fcollonval — COMMENTED** [review]
Thanks a lot @blink1073 for this - I have a question about the merge action.

File: jupyter_releaser/lib.py
Comment (blink1073) [reply jupyter_releaser/lib.py:186]: Good question, I'll test it out on my test repo.

Comment (blink1073) [reply jupyter_releaser/lib.py:186]:
```
====Error Body====
{
  "message": "Required status check \"test\" is expected.",
  "documentation_url": "https://docs.github.com/articles/about-protected-branches"
}
```

Hmm, given that there is a way to do this manually on a PR, I'll have to look at the GitHub API docs.

Comment (blink1073) [reply jupyter_releaser/lib.py:186]: Okay, I think a combination of using `[ci skip]` and then a backoff until the PR can be merged might work.

**Comment by blink1073** [comment]
Okay, in order to make this work we need to enable fine-grained access tokens on the orgs.  These access tokens will only have "Read and write" "Pull Request" permission on a particular repository.  We use that token only for creating PRs, since otherwise the workflows don't run on the PRs.  We don't need to store them anywhere other then the "Environment Secrets".  The net overall benefit is that we don't need any more bot accounts for GitHub, and anyone one of the admins on the org can manage and update these tokens.  The tokens themselves are much more scoped down than the admin tokens we were forced to use before.  The owner of the token will regenerate the token and update the Environment secret as needed.  The one downside I see is that the PRs themselves are in the name of whoever created the token, but the commits will still be in the name of whichever admin runs the deployment.

image

image


image

image

**Comment by fcollonval** [comment]
Thanks a lot @blink1073 this is a nice improvement; the new token required will indeed be far better.

**Comment by jtpio** [comment]
> The one downside I see is that the PRs themselves are in the name of whoever created the token, but the commits will still be in the name of whichever admin runs the deployment.

Hmm in that case maybe we should keep using the bots, so the PRs are opened from the bot account, so it's more "neutral"? Otherwise it would be a bit strange if the GitHub account used for opening the PR does not correspond to the person doing the release.

**Comment by blink1073** [comment]
We would need to give the bots admin access to the org itself:

image

**Comment by jtpio** [comment]
Otherwise a maintainer who wants to make a new release would have to go to the repository settings and put their token. Then after the release remove the secret. The next maintainer would also have to use their token to make a release.

The downside is that this adds a couple more clicks and burden for making a release, but should help associate the PR with the maintainer the release.

**Comment by jtpio** [comment]
Or maybe the workflow dispatch could take the user token as input as a mandatory field. Although not sure this can be done securely without potentially leaking the token.

**Comment by blink1073** [comment]
That could be up to the project.  Personally, I don't think it would be worth it, since the commits are still in the workflow runner's username.

**Comment by fcollonval** [comment]
> Personally, I don't think it would be worth it, since the commits are still in the workflow runner's username.

:+1:

**Comment by blink1073** [comment]
This commit has the docs and example updates: [`022cba1` (#521)](https://github.com/jupyter-server/jupyter_releaser/pull/521/commits/022cba1ff06ae05958f08e0c9aa2490236a9dfd3)

**Review by fcollonval — APPROVED** [review]
Thanks @blink1073

**Review by jtpio — APPROVED** [review]
Thanks!

**Comment by blink1073** [comment]
@fcollonval @jtpio I have an idea: instead of using the personal access token, we can add an option to `jupyterlab-probot` to close and reopen a PR meeting certain criteria.   It already subscribes to Pull Requests and it has access to open/close them.  This would trigger the workflows to run, and would require no ongoing maintenance.  We can still make the PAT available for folks who don't want to use `jupyterlab-probot` but still have merge requirements.

**Comment by jtpio** [comment]
> we can add an option to `jupyterlab-probot` to close and reopen a PR meeting certain criteria

This would also be useful to trigger CI after Playwright snapshots updates: https://github.com/jupyterlab/jupyterlab/issues/13505

**Comment by blink1073** [comment]
Yeah I like that idea.  The alternative would be to add a PAT to the [workflow](https://github.com/jupyterlab/jupyterlab/blob/main/.github/workflows/galata-update.yml), but I don't think we want to start doing that in general, because it would be easy to introduce a security hole.

**Comment by blink1073** [comment]
Now that that the PR bot can close/reopen, the next step is to add that as an option, and make the comment if it is given, and update the instructions accordingly.

**Comment by blink1073** [comment]
I think this actually conflicts with the ability to have silent releases, since we'd have to have a PR instead of a push.  

New idea: use the user's `GITHUB_TOKEN`, and recommend the use of an environment.  Technically our check for the github actor's permissions would still work, but environment is still better.

To handle forwardport and silent changelog PRs, we add an option to make the PR comment to toggle the PR.

**Pull request closed without merging**
